packages/i18n/src/index.ts

```diff
@@ -157,10 +158,135 @@ export const createI18nContext = (
+  ? string
+  : T extends (...args: infer A) => unknown
+  ? T & { __length: A["length"] }
+  : {
+      [K in keyof T]: AddFunctionLengths<T[K]>;
+    };
+
+type RemoveFunctionLengths<T> = T extends string
+  ? string
+  : T extends (args: infer A) => infer R
+  ? (args: A) => R
+  : {
+      [K in keyof T]: RemoveFunctionLengths<T[K]>;
+    };
+
+type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void
+  ? I
+  : never;
+
+type GuaranteeIdenticalSignatures<T extends Dictionaries<I18nObject>> = Record<
+  keyof T,
+  RemoveFunctionLengths<UnionToIntersection<AddFunctionLengths<T[keyof T]>>>
+>;
+
+const buildI18nChain = <T extends I18nObject>(obj: T): I18nPath<T> => {
+  const keys = Object.keys(obj) as (keyof T)[];
+  const paths = keys.reduce((acc, key) => {
+    const value = obj[key] as any;
+    if (typeof value === "object") {
+      return {
+        ...acc,
+        [key]: buildI18nChain(value)
+      } as any;
+    } else if (typeof value === "function") {
+      return {
+        ...acc,
+        [key]: obj[key]
+      };
+    }
+    if (typeof value === "string") {
+      return {
+        ...acc,
+        [key]: (args: I18nFormatOptions) => {
+          return value.replace(/{{(.*?)}}/g, (_, key) => deepReadObject(args, key, ""));
+        }
+      };
+    } else {
+      throw new TypeError(
+        `Unsupported data format on the keys. Values must resolve to a string or a function that returns a string. Key name: "${
+          key as string
+        }"`
+      );
+    }
+  }, {} as Partial<I18nPath<T>>);
+
+  return paths as I18nPath<T>;
+};
+
+export interface I18nDictionary {
+  readonly [x: string]: string | ((...args: any) => string) | I18nDictionary;
+}
+
+export const makeChainedI18nContext = <
+  T extends Dictionaries<I18nObject>,
+  K extends keyof T
+>(props: {
+  dictionaries: T & GuaranteeIdenticalSignatures<T>;
+  locale: keyof T;
+  setContext?: boolean;
+}) => {
+  const [locale, _setLocale] = createSignal<K>(props.locale as K);
+  const [translate, _setTranslate] = createStore(buildI18nChain(props.dictionaries[props.locale]));
+
+  const utils = {
+    locale,
+    setLocale(newLocale: K): K {
+      _setTranslate(buildI18nChain(props.dictionaries[newLocale]));
+      _setLocale(() => newLocale);
+      return newLocale;
+    },
+    getDictionary(language?: K): T[K] {
+      if (language) return props.dictionaries[language];
+      return props.dictionaries[locale()];
+    }
+  };
+
+  const chainedI18nContext = createContext<[typeof translate, typeof utils] | null>(
+    props.setContext ? [translate, utils] : null
+  );
+
+  return {
+    I18nProvider: (props: { children: JSXElement }): JSXElement => {
+      return createComponent(chainedI18nContext.Provider, {
+        value: [translate, utils],
+        get children() {
+          return props.children;
+        }
+      });
+    },
+    useI18nContext: () => useContext(chainedI18nContext)
+  };
```

Comment:
A primitive for i18n should worry only about i18n. How do you decide to share returned signals with the rest of the app? that's a separate concern. It could be done in many ways. What if you prefer to have one giant context provider for one app instead of multiple smaller ones? What if you want to share the translations by just exporting them from a file?
What do you think about separating the current primitive into two?
- one that creates the translations object and manages the locales. e.g. `createChainedI18n`
- and one that could wrap it into a convenient context if the user wants it? e.g. `createI18nContext`
Although I'm not sure if the second one for context is even necessary. If it could be abstracted and added to the /context package to be used with any state it would be much better than creating one for each.

Reply:
Okay. Maybe it's better to make this just a utility for i18n that people can use to build out chained dictionaries then they decide how to ingest it! Would make it a lot less opinionated and a smaller bundle.

Reply:
After playing with this more, I believe that i18n needs context. I split it up, but without providing a friendly way to interact with the translations, a user will need to create their own wrapper every time to use the data and handle locales.